Update Working with Dependencies -> Dependency Resolution Section (#30928)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/dep-man/dependency-resolution/artifact_transforms.adoc

[[artifact_transforms]]
= Artifact Transforms

What if you want to makes changes to the files contained in one of your dependencies before you use it?

For example, you might want to unzip a compressed file, adjust the contents of a JAR, or delete unnecessary files from a dependency that contains multiple files prior to using the result in a task.

Gradle has a built-in feature for this called **Artifact Transforms**.
With Artifact Transforms, you can modify, add to, remove from the set files (or artifacts) - like JAR files - contained in a dependency.

> This is done as the last step in resolving a dependency, before tasks or tools like the IDE uses it.

Reviewer: ```suggestion
This is done as the last step when resolving artifacts, before tasks or tools like the IDE can consume the artifacts.
```

== Artifact Transforms Overview

Each component exposes a set of <<declaring_dependencies_adv.adoc#sec:resolvable-consumable-configs,**variants**>>, where each variant is identified by a set of <<variant_model.adoc#sec:abm_configuration_attributes,**attributes**>> (i.e., key-value pairs such as `debug=true`).

When Gradle resolves a configuration, it looks at each dependency, resolves it to a component, and selects the corresponding variant from that component that matches the requested attributes.

> If the component does not have a matching variant, resolution fails *unless* Gradle can construct a sequence of transformations that will modify an existing variant to create a valid match (without changing its transitive dependencies).

Reviewer: ```suggestion
If the component does not have a matching variant, resolution fails *unless* Gradle can construct a sequence of transformations that will modify an existing artifact to create a valid match (without changing its transitive dependencies).
```

1. **Implement a Transform**: You define an artifact transform by creating a class that implements the link:{javadocPath}/org/gradle/api/artifacts/transform/TransformAction.html[`TransformAction`] interface.
This class specifies how the input artifact should be transformed into the output artifact.

2. **Declare request Attributes**: Attributes (key-value pairs used to describe different variants of a component) like `org.gradle.usage=java-api` and `org.gradle.usage=java-runtime` are used to specify the desired artifact format or type.

3. **Register a Transform**: You register the transform by using the link:{javadocPath}/org/gradle/api/artifacts/dsl/DependencyHandler.html#registerTransform(java.lang.Class,org.gradle.api.Action)[`registerTransform()`] method of the link:{javadocPath}/org/gradle/api/Project.html#dependencies(groovy.lang.Closure)[`dependencies`] block.
This method tells Gradle that a transform can be used to modify the artifacts of any variant that possesses the given "from" attributes.
It also tells Gradle what new set of "to" attributes will describe the format or type of the resulting artifacts.

> 4. **Use the Transform**: When a resolution requires a variant that isn't already present in the selected component (because none of the actual variants possess compatible attributes to the requested attributes), Gradle doesn't just give up!

Reviewer: ```suggestion
4. **Use the Transform**: When a resolution requires an artifact that isn't already present in the selected component (because none of the actual artifact possess compatible attributes to the requested attributes), Gradle doesn't just give up!
```

include::sample[dir="snippets/dependencyManagement/artifactTransforms-simple/groovy",files="build.gradle[tags=artifact-transform-registration]"]
====

> "To" attributes are used to supply describe the format or type of the artifacts that this transform can use as an input, and "from" attributes to describe the format or type of the artifacts that it produces as an output.

Reviewer: ```suggestion
"To" attributes are used to describe the format or type of the artifacts that this transform can use as an input, and "from" attributes to describe the format or type of the artifacts that it produces as an output.
```

=== 4. Use the Transform

During a build, Gradle *automatically* runs registered transforms to satisfy a resolution request if a match is not directly available.

> Since no variants exist supplying the requested format (as none contain the `artifactType` attribute with a value of `"transformed-jar"`), Gradle attempts to construct a chain of transformations that will supply it.

Reviewer: ```suggestion
Since no variants exist supplying artifacts of requested format (as none contain the `artifactType` attribute with a value of `"transformed-jar"`), Gradle attempts to construct a chain of transformations that will supply it.
```

Gradle's search finds `MyTransform`, which is registered as producing the requested format, so it will automatically be run.
Running this transform action modifies the artifacts of an existing source variant to produce new artifacts that are delivered to the consumer, in the requested format.

> Gradle produces a "virtual variant" of the component as part of this process.

Reviewer: ```suggestion
Gradle produces a "virtual artifact set" of the component as part of this process.
```

[[sec:abm_artifact_transforms]]
== Understanding Artifact Transforms

Dependencies can have different <<declaring_dependencies_adv.adoc#sec:resolvable-consumable-configs,**variants**>>, essentially different versions or forms of the same dependency.

> These variants can each provide a different collections of files, meant to satisfy different use cases, such as compiling code, browsing documentation or running applications.

Reviewer: ```suggestion
These variants can each provide a different artifact set, meant to satisfy different use cases, such as compiling code, browsing documentation or running applications.
```

For the project dependency, the `java-api,classes` and the `java-api,jars` variants have the same transitive dependencies but different artifacts — the `classes` directories and the `JAR` files respectively.

When Gradle resolves a configuration, it uses the attributes defined to select the appropriate variant of each dependency.
The attributes that Gradle uses to determine which variant to select are called the **requested attributes**.

For example, if a configuration requests `org.gradle.usage=java-api` and `org.gradle.libraryelements=classes`, Gradle will select the variant of each dependency that matches these attributes (in this case, classes directories intended for use as an API during compilation).
Matches do not have to exact, as some attribute values can be identified to Gradle as *compatible* with other values and used interchangeably during
<<variant_attributes.adoc#sec:abm_compatibility_rules,**matching**>>.

Sometimes, a dependency might not have a variant with attributes that match the requested attributes.

> In such cases, Gradle can transform one variant into another "vitual variant" by modifying its artifacts without changing its transitive dependencies.

Reviewer: ```suggestion
In such cases, Gradle can transform one variant's artifacts into another "virtual artifact set" by modifying its artifacts without changing its transitive dependencies.
```

IMPORTANT: Gradle will not attempt to select or run Artifact Transforms when a variant of the dependency matching the requested attributes already exists.

For example, if the requested variant is `java-api,classes`, but the dependency only has `java-api,jar`, Gradle can potentially transform the `JAR` file into a `classes` directory by unzipping it using an Artifact Transform that is registered with these attributes.

== Understanding Artifact Transforms Chains

When Gradle resolves a configuration and a dependency does not have a variant with the requested attributes, it attempts to find a chain of one or more Artifact Transforms that can be run sequentially to create the desired variant.
This process is called **Artifact Transform selection**:

image::artifact-transform-5.png[]

=== The Artifact Transform Selection Process:

1. **Start with requested Attributes**:

> - Gradle starts with the attributes specified in a dependency resolution request (typically those present on the configuration being resolved).

Reviewer: ```suggestion
- Gradle starts with the attributes specified on the configuration being resolved, appends any attributes specified on an `ArtifactView`, and finally appends any attributes declared directly on the dependency. 
```
